### PR TITLE
Add PyTorch 2.0 Flash Attention

### PR DESCRIPTION
**Pull request by conceptofmind**
Hi Phil,

This PR adds support for Flash Attention in Triton. Tri Dao had given me some input on how to properly handle the tensor shapes for multi-query single-key-value attention with Flash Attention.

I was unsure which version of Flash Attention you wanted to add. I can work on changing it to the CUDA version if that is preferred.

```python
import torch
from palm_rlhf_pytorch import PaLM

palm = PaLM(
    num_tokens = 20000,
    dim = 512,
    depth = 12,
    flash_attn=True,
).half().cuda()

seq = torch.randint(0, 20000, (1, 2048)).cuda()

loss = palm(seq, return_loss = True)
print(loss)
```
Thank you,

Enrico

### CODE REVIEW

**Comment by conceptofmind** [comment]
I also was not sure whether you want the attention mechanisms separated into two different functions or if just an `if/else` statement would be sufficient.

**Comment by lucidrains** [comment]
@conceptofmind oh thanks for doing this! so actually my plan had been to just wait for pytorch 2.0 official release

flash attention will be available as a part of the functional package, probably the same code that was used to train llama

hopefully it is soon..

**Comment by conceptofmind** [comment]
> @conceptofmind oh thanks for doing this! so actually my plan had been to just wait for pytorch 2.0 official release
> 
> flash attention will be available as a part of the functional package, probably the same code that was used to train llama
> 
> hopefully it is soon..

Ok. Awesome to know that it will be directly integrated with PyTorch now! Do you want me to close this PR or leave it open as a reference for others until they update the new 2.0 release?

**Comment by lucidrains** [comment]
@conceptofmind you can leave it open, as pytorch 2.0 just got released this week with [flash attention support](https://deploy-preview-1313--pytorch-dot-org-preview.netlify.app/blog/pytorch-2.0-release/)! i'll probably try this weekend to upgrade to cuda 11.7 and to pytorch 2.0 and see if i can get it integrated

but if you want to give it a try, i would definitely welcome the contribution!

**Comment by conceptofmind** [comment]
@lucidrains I updated to use the PyTorch 2.0 version of Flash Attention.

Following the documentation:
- https://pytorch.org/docs/master/generated/torch.nn.functional.scaled_dot_product_attention.html?highlight=scaled_dot_product#torch.nn.functional.scaled_dot_product_attention
- https://pytorch.org/docs/master/backends.html#torch.backends.cuda.enable_math_sdp
```python
torch.backends.cuda.sdp_kernel(enable_flash=True, enable_math=False, enable_mem_efficient=False):
    F.scaled_dot_product_attention(query,key,value)
```

Confirmed with Patrick from PyTorch: https://discuss.pytorch.org/t/flash-attention/174955

It trained locally for me with CUDA 11.7 on an RTX 3090. I would just do a sanity check though since it is still so new.

Thank you,

Enrico

**Comment by lucidrains** [comment]
@conceptofmind wow, thank you Enrico! this looks great :pray: 

do you want to add a citation for flash attention, and also do a version check to make sure pytorch is 2.0 before flash attention can be enabled?

**Comment by lucidrains** [comment]
@conceptofmind haha, i'm upgrading to pytorch 2.0 this morning, wish me luck (need to upgrade my CUDA from 11.4 too)

File: palm_rlhf_pytorch/palm.py [hunk 1]
```diff
@@ -237,30 +240,56 @@ def forward(
...
+            # pytorch 2.0 flash attn: q, k, v, mask, dropout, causal, softmax_scale
+            
+            with torch.backends.cuda.sdp_kernel(
+                enable_flash=self.flash_attn, 
+                enable_math=False, 
+                enable_mem_efficient=False
+            ):
+                out = F.scaled_dot_product_attention(
+                    q, k, v,
+                    attn_mask = mask,
```
[the remark this answers is not in the document]
Comment (lucidrains) [reply palm_rlhf_pytorch/palm.py:262]: their team should add this though, it is pretty standard these days

**Comment by lucidrains** [comment]
@conceptofmind one more thing to test is whether their context can naturally fallback to cpu, if cuda is not found

**Comment by conceptofmind** [comment]
> @conceptofmind wow, thank you Enrico! this looks great pray
> 
> do you want to add a citation for flash attention, and also do a version check to make sure pytorch is 2.0 before flash attention can be enabled?

I will add the citation for Flash Attention. Is this what you imagined for an assertion:
```python
if self.flash_attn:
  try:
    assert torch.__version__ >= "2.0.0"
  except:
    raise Exception("flash attention requires pytorch 2.0")
```

> @conceptofmind haha, i'm upgrading to pytorch 2.0 this morning, wish me luck (need to upgrade my CUDA from 11.4 too)

I upgraded from 11.3 to 11.7. Always fun installing new CUDA versions and having to scan through 10s of files to make sure there are no residual conflicting artifacts from previous installations :laughing:

> @conceptofmind one more thing to test is whether their context can naturally fallback to cpu, if cuda is not found

I believe the way the context manager is set up is that it checks what is available from the three and defaults to the best option. I had hardcoded it to use FlashAttention to ensure that is was used. I can leave it to the default:
```python
torch.backends.cuda.sdp_kernel(enable_flash=True, enable_math=True, enable_mem_efficient=True):
    F.scaled_dot_product_attention(query,key,value)
```
Or I can add some sort of filter if an A100 is available:
```python
if enable_flash == True:
    enable_math = False
    enable_mem_efficient = False
else:
    enable_math = True
    enable_mem_efficient = True
```
One thing to note is that PyTorch 2.0 Flash Attention seems to throw a kernel error if a compatible GPU architecture is not available.
```bash
RuntimeError: No available kernel.  Aborting execution.
```
Should we include something that checks for an A100:
```python
try:
    if torch.cuda.is_available():
        device_properties = torch.cuda.get_device_properties(0)
        if device_properties.major == 8 and device_properties.minor == 0:
            print('The A100 architecture is available.')
        else:
            print('The A100 architecture is not available.')
    else:
        print('No GPU is available.')
except RuntimeError as error:
    print(f'An error occurred: {error}.')
```
Let me know what you think is best!

Thank you,

Enrico

File: palm_rlhf_pytorch/palm.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (conceptofmind) [reply palm_rlhf_pytorch/palm.py:262]: I will do this for the mask:
```python
# Check if mask exists and expand to the compatible shape
if exists(mask):
    mask = mask.unsqueeze(1).expand_as(q)
```

**Comment by lucidrains** [comment]
yup exactly! but you would do `from packaging import version`, and then `version.parse(torch.__version__) >= version.parse('2.0.0')`, as string comparison would have a different behavior.

and yes, try catches would be perfect for if the kernel selection heuristic fails

**Comment by conceptofmind** [comment]
@lucidrains 

> yup exactly! but you would do `from packaging import version`, and then `version.parse(torch.__version__) >= version.parse('2.0.0')`, as string comparison would have a different behavior.
> 
> and yes, try catches would be perfect for if the kernel selection heuristic fails

Well I think the kernel selection works. :laughing:
```python
# Check if there is a compatible device for flash attention

try:
    flash_device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    if flash_device.type == 'cuda':
        device_properties = torch.cuda.get_device_properties(device)
        if device_properties.major == 8 and device_properties.minor == 0:
            print('A100 GPU detected, using flash attention')
            enable_flash = True
            enable_math = False
            enable_mem_efficient = False
        else:
            print('Non-A100 GPU detected, using either math or mem efficient attention')
            enable_flash = False
            enable_math = True
            enable_mem_efficient = True
    else:
        # Default context manager settings with CPU
        print('CPU detected, using default context manager settings')
        enable_flash = True
        enable_math = True
        enable_mem_efficient = True
except RuntimeError as error:
    print(f'An error occurred: {error}.')
```
Pushed the other updates as well.

Hopefully, everything looks better now.

Thank you,

Enrico

**Comment by lucidrains** [comment]
@conceptofmind yes, almost there! left one comment

File: palm_rlhf_pytorch/palm.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (conceptofmind) [reply palm_rlhf_pytorch/palm.py:300]: My bad. I read that as quite literally doing the same thing as for k and v. I will change it now.

File: palm_rlhf_pytorch/palm.py [hunk 2]
```diff
@@ -237,30 +241,93 @@ def forward(
...
+
+            # Check if mask exists and expand to compatible shape
+
+            if exists(mask):
+                mask = mask.unsqueeze(1).expand_as(q)
+
+            # Check if there is a compatible device for flash attention
+
+            try:
+                flash_device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
```
Comment (lucidrains) [inline palm_rlhf_pytorch/palm.py:269]: oh, actually this logic should also depend on whether the model is on cuda or not, as even if cuda is available, someone could run it on cpu

File: palm_rlhf_pytorch/palm.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (lucidrains) [reply palm_rlhf_pytorch/palm.py:300]: no worries! yea, masking is the most confusing part of transformers

File: palm_rlhf_pytorch/palm.py (hunk 2 above)
Comment (conceptofmind) [reply palm_rlhf_pytorch/palm.py:269]: @lucidrains 

Do you mean something more along the lines of this:
```python
if torch.cuda.is_available():
    if x.device.type == 'cuda':
        flash_device = torch.device('cuda')
    else:
        flash_device = torch.device('cpu')
else:
    flash_device = torch.device('cpu')
```

Comment (lucidrains) [reply palm_rlhf_pytorch/palm.py:269]: now that i thought about it some more, not sure what the context manager is doing

i think your logic may be fine!

Comment (lucidrains) [reply palm_rlhf_pytorch/palm.py:269]: > Do you mean something more along the lines of this:
> 
> ```python
> if torch.cuda.is_available():
>     if x.device.type == 'cuda':
>         flash_device = torch.device('cuda')
>     else:
>         flash_device = torch.device('cpu')
> else:
>     flash_device = torch.device('cpu')
> ```

yup something like this, but i don't know if the context manager is able to fallback naturally

i'll figure it out by evening!

Comment (conceptofmind) [reply palm_rlhf_pytorch/palm.py:269]: > This context manager can be used to temporarily enable or disable any of the three backends for scaled dot product attention. Upon exiting the context manager, the previous state of the flags will be restored.
> 
> All implementations are enabled by default. Scaled dot product attention attempts to automatically select the most optimal implementation based on the inputs.
> 
> PyTorch 2.0 will automatically select the optimal implementation for your use case, but you can also toggle them individually for finer-grained control.

These are the parts of the documentation I saw related to the context manager. It seems like it is supposed to automatically default to whatever is available.

Comment (lucidrains) [reply palm_rlhf_pytorch/palm.py:269]: ok, that sounds great!

File: palm_rlhf_pytorch/palm.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (conceptofmind) [reply palm_rlhf_pytorch/palm.py:300]: So the mask is currently `b j` but to be compatible with PyTorch 2.0 masking it should be `b ... j i`.

[the remark this answers is not in the document]
Comment (lucidrains) [reply palm_rlhf_pytorch/palm.py:300]: but you should run a quick forward to check if it works or not

**Comment by conceptofmind** [comment]
Updated to include your recommendation for masking and additional logic.

I tried training and ran three forward passes. The loss went down.

No Flash
```
import torch
from palm_rlhf_pytorch import PaLM

palm = PaLM(
    num_tokens = 20000,
    dim = 512,
    depth = 12,
    flash_attn=False,
).cuda()

seq = torch.randint(0, 20000, (1, 2048)).cuda()

loss = palm(seq, return_loss = True)
loss.backward()
print(loss)
```
Flash
```
import torch
from palm_rlhf_pytorch import PaLM

palm = PaLM(
    num_tokens = 20000,
    dim = 512,
    depth = 12,
    flash_attn=True,
).cuda()

seq = torch.randint(0, 20000, (1, 2048)).cuda()

loss = palm(seq, return_loss = True)
loss.backward()
print(loss)
```
CPU
```
import torch
from palm_rlhf_pytorch import PaLM

palm = PaLM(
    num_tokens = 20000,
    dim = 512,
    depth = 12,
)

seq = torch.randint(0, 20000, (1, 2048))

loss = palm(seq, return_loss = True)
loss.backward()
print(loss)
```

File: palm_rlhf_pytorch/palm.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (conceptofmind) [reply palm_rlhf_pytorch/palm.py:308]: I will scold them for you. Then lament at the difficulty of figuring out attention masks.

Joking aside. I appreciate your help and input. It is always greatly educational.

**Comment by lucidrains** [comment]
Nice! Want to also test it with non-causal with mask?

**Comment by conceptofmind** [comment]
> Nice! Want to also test it with non-causal with key padding mask?

Will start writing some tests for non-causal now.

**Comment by conceptofmind** [comment]
I tried with these configurations for `causal=False`:

No Flash
```python
import torch
from palm_rlhf_pytorch import PaLM

palm = PaLM(
    num_tokens = 20000,
    dim = 512,
    depth = 12,
    causal = False,
    flash_attn = False,
)

seq = torch.randint(0, 20000, (1, 2048))

loss = palm(seq, return_loss = True)
loss.backward()
print(loss)
```
Flash
```python
import torch
from palm_rlhf_pytorch import PaLM

palm = PaLM(
    num_tokens = 20000,
    dim = 512,
    depth = 12,
    causal = False,
    flash_attn = True,
)

seq = torch.randint(0, 20000, (1, 2048))

loss = palm(seq, return_loss = True)
loss.backward()
print(loss)
```
Flash
```python
import torch
from palm_rlhf_pytorch import PaLM, RewardModel

palm = PaLM(
    num_tokens = 20000,
    dim = 512,
    depth = 12,
    causal = False,
    flash_attn = True,
)

reward_model = RewardModel(
    palm,
    num_binned_output = 5 # say rating from 1 to 5
).cuda()

# mock data

seq = torch.randint(0, 20000, (1, 1024)).cuda()
prompt_mask = torch.zeros(1, 1024).bool().cuda() # which part of the sequence is prompt, which part is response
labels = torch.randint(0, 5, (1,)).cuda()

# train

loss = reward_model(seq, prompt_mask = prompt_mask, labels = labels)
loss.backward()
```
No Flash
```python
import torch
from palm_rlhf_pytorch import PaLM, RewardModel

palm = PaLM(
    num_tokens = 20000,
    dim = 512,
    depth = 12,
    causal = False,
    flash_attn = False,
)

reward_model = RewardModel(
    palm,
    num_binned_output = 5 # say rating from 1 to 5
).cuda()

# mock data

seq = torch.randint(0, 20000, (1, 1024)).cuda()
prompt_mask = torch.zeros(1, 1024).bool().cuda() # which part of the sequence is prompt, which part is response
labels = torch.randint(0, 5, (1,)).cuda()

# train

loss = reward_model(seq, prompt_mask = prompt_mask, labels = labels)
loss.backward()
print(loss)
```
I used the PyTorch version for expanding the mask you provided to keep it consistent.

Thank you,

Enrico

**Comment by lucidrains** [comment]
@conceptofmind graciously accept! thank you for all your work on this PR 🙏

**Pull request merged**